### PR TITLE
new data_source and resource slb_domain_extensions

### PR DESCRIPTION
**Pull request by xujunnnn**
=== RUN   TestAccAlicloudSlbDomainExtensionBasic
--- PASS: TestAccAlicloudSlbDomainExtensionBasic (15.33s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     15.397s

=== RUN   TestAccAlicloudSlbDomainExtensionsDataSource_basic
--- PASS: TestAccAlicloudSlbDomainExtensionsDataSource_basic (21.35s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     21.417s

### CODE REVIEW

File: alicloud/data_source_alicloud_slb_domain_extensions.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		for _, vv := range v.([]interface{}) {
+			idsMap[Trim(vv.(string))] = Trim(vv.(string))
+		}
+	}
+
+	raw, err := client.WithSlbClient(func(slbClient *slb.Client) (interface{}, error) {
+		return slbClient.DescribeDomainExtensions(args)
+	})
+	if err != nil {
+		return fmt.Errorf("DescribeDomainExtensions got an error: %#v", err)
```
Comment (Lexsss) [inline alicloud/data_source_alicloud_slb_domain_extensions.go]: return WrapErrorf(err, DataDefaultErrorMsg, "alicloud_slb_domain_extensions", request.GetActionName(), AlibabaCloudSdkGoERROR)

File: alicloud/data_source_alicloud_slb_domain_extensions.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	}
```
Comment (Lexsss) [inline alicloud/data_source_alicloud_slb_domain_extensions.go]: addDebug(request.GetActionName(), raw, request.RpcRequest, request)

File: alicloud/data_source_alicloud_slb_domain_extensions.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	resp, _ := raw.(*slb.DescribeDomainExtensionsResponse)
+	if resp == nil {
```
Comment (Lexsss) [inline alicloud/data_source_alicloud_slb_domain_extensions.go]: no need nil check

File: alicloud/data_source_alicloud_slb_domain_extensions.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+			"id":                    domainExtension.DomainExtensionId,
+			"domain":                domainExtension.Domain,
+			"server_certificate_id": domainExtension.ServerCertificateId,
+		}
+		ids = append(ids, domainExtension.DomainExtensionId)
+		s = append(s, mapping)
+	}
+	d.SetId(dataResourceIdHash(ids))
+	if err := d.Set("slb_domain_extensions", s); err != nil {
+		return err
```
Comment (Lexsss) [inline alicloud/data_source_alicloud_slb_domain_extensions.go]: WrapError(err)

File: alicloud/data_source_alicloud_slb_domain_extensions_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+func TestAccAlicloudSlbDomainExtensionsDataSource_basic(t *testing.T) {
+	rand := acctest.RandInt()
+	basicConf := dataSourceTestAccConfig{
+		existConfig: testAccCheckAlicloudDomainExtensionDataSourceConfig(rand, map[string]string{
+			"load_balancer_id": `"${alicloud_slb_domain_extension.default.load_balancer_id}"`,
+			"frontend_port":    `"${alicloud_slb_domain_extension.default.frontend_port}"`,
+		}),
+	}
+	idsConf := dataSourceTestAccConfig{
```
Comment (Lexsss) [inline alicloud/data_source_alicloud_slb_domain_extensions_test.go]: remove this step

File: alicloud/data_source_alicloud_slb_domain_extensions.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+				},
+			},
+		},
+	}
+}
+
+func dataSourceAlicloudSlbDomainExtensionsRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	args := slb.CreateDescribeDomainExtensionsRequest()
```
Comment (Lexsss) [inline alicloud/data_source_alicloud_slb_domain_extensions.go]: please rename `args` to `request`.

File: alicloud/data_source_alicloud_slb_domain_extensions.go (hunk 3 above)
Comment (Lexsss) [inline alicloud/data_source_alicloud_slb_domain_extensions.go]: please rename `resp` to `response`.

File: alicloud/resource_alicloud_slb_domain_extension.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceAliyunSlbDomainExtensionCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	req := slb.CreateCreateDomainExtensionRequest()
```
Comment (Lexsss) [inline alicloud/resource_alicloud_slb_domain_extension.go]: rename

File: alicloud/resource_alicloud_slb_domain_extension.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	req.LoadBalancerId = d.Get("load_balancer_id").(string)
+	req.ListenerPort = requests.NewInteger(d.Get("frontend_port").(int))
+	req.Domain = d.Get("domain").(string)
+	req.ServerCertificateId = d.Get("server_certificate_id").(string)
+	err := resource.Retry(3*time.Minute, func() *resource.RetryError {
```
Comment (Lexsss) [inline alicloud/resource_alicloud_slb_domain_extension.go]: Please refer to the retry in the `resource_alicloud_cen_route_entry.go`, modified the way to return the error and addDebug/SetId.

[https://github.com/terraform-providers/terraform-provider-alicloud/blob/master/alicloud/resource_alicloud_cen_route_entry.go](url)

File: alicloud/resource_alicloud_slb_domain_extension.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	}
+	d.Partial(false)
+	return resourceAliyunSlbDomainExtensionRead(d, meta)
+}
+
+func resourceAliyunSlbDomainExtensionDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	req := slb.CreateDeleteDomainExtensionRequest()
+	req.DomainExtensionId = d.Id()
+	return resource.Retry(5*time.Minute, func() *resource.RetryError {
```
Comment (Lexsss) [inline alicloud/resource_alicloud_slb_domain_extension.go]: Only `DeleteDomainExtension` in this retry, refer to the Delete in the `resource_alicloud_cen_route_entry.go`.

File: alicloud/resource_alicloud_slb_domain_extension.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+		client := meta.(*connectivity.AliyunClient)
+		slbService := SlbService{client}
+		if _, err := slbService.DescribeDomainExtensionAttribute(d.Id()); err != nil {
+			if NotFoundError(err) {
+				return nil
+			}
+			return resource.NonRetryableError(WrapError(err))
+		}
+		return resource.RetryableError(WrapErrorf(err, DeleteTimeoutMsg, d.Id(), req.GetActionName(), ProviderERROR))
+	})
```
Comment (Lexsss) [inline alicloud/resource_alicloud_slb_domain_extension.go]: If this API is called asynchronously, please add the WaitFor method.

File: alicloud/resource_alicloud_slb_domain_extension_test.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+					"server_certificate_id": "${alicloud_slb_server_certificate.new.id}",
+				}),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheck(map[string]string{
+						"load_balancer_id":      CHECKSET,
+						"frontend_port":         CHECKSET,
+						"domain":                "www.test.com",
+						"server_certificate_id": CHECKSET,
+					}),
+				),
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_slb_domain_extension_test.go]: where is `Importer` test step?

File: vendor/vendor.json [hunk 12]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (xiaozhu36) [inline vendor/vendor.json]: please use go module

File: alicloud/data_source_alicloud_slb_domain_extensions.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+			},
+			"frontend_port": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"output_file": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"domain_extensions": {
```
Comment (xiaozhu36) [inline alicloud/data_source_alicloud_slb_domain_extensions.go]: please use `extensions` instead `domain_extensions `

File: alicloud/resource_alicloud_slb_domain_extension.go [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+	d.Set("server_certificate_id", domainExtension.ServerCertificateId)
+	d.Set("frontend_port", domainExtension.ListenerPort)
+
+	return nil
+}
+
+func resourceAliyunSlbDomainExtensionUpdate(d *schema.ResourceData, meta interface{}) error {
+	d.Partial(true)
+
+	if d.HasChange("server_certificate_id") {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_slb_domain_extension.go]: If there is only one api operation, the `partial action` is redundant.

File: website/docs/d/slb_domain_extension.html.markdown [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+sidebar_current: "docs-alicloud-resource-slb-domain-extension"
+description: |-
+  Provides a Load Banlancer domain extension Resource and add it to one Listener.
+---
+
+# alicloud\_slb\_domain_extension
+
+This data source provides the domain extensions associated with a server load balancer listener.
+
+-> **NOTE:** Available in 1.59.0+
```
Comment (xiaozhu36) [inline website/docs/d/slb_domain_extension.html.markdown]: 1.60.0+

File: website/docs/r/slb_domain_extension.html.markdown [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+description: |-
+  Provides a Load Banlancer domain extension Resource and add it to one Listener.
+---
+
+# alicloud\_slb\_domain_extension
+
+HTTPS listeners of guaranteed-performance SLB support configuring multiple certificates, allowing you to forward requests with different domain names to different backend servers.
+Please refer to the [documentation](https://www.alibabacloud.com/help/doc-detail/85956.htm?spm=a2c63.p38356.b99.40.1c881563Co8p6w) for details.
+
+-> **NOTE:** Available in 1.59.0+
```
Comment (xiaozhu36) [inline website/docs/r/slb_domain_extension.html.markdown]: 1.60.0

File: website/docs/d/slb_domain_extension.html.markdown [hunk 17]
```diff
@@ -0,0 +1,40 @@
+---
```
Comment (xiaozhu36) [inline website/docs/d/slb_domain_extension.html.markdown]: please register the docs in the `alicloud.erb`

**Pull request merged**